jujugui/static/gui/src/app/components/post-deployment/post-deployment.js

```diff
@@ -0,0 +1,204 @@
+/* Copyright (C) 2015 Canonical Ltd. */
+
+'use strict';
+
+/**
+  Displays post-deployment information for usage and quickstart.
+  Shown when a charm or bundle is added to the canvas, and on deploy.
+*/
+class PostDeployment extends React.Component {
+  constructor(props) {
+    super(props);
+    this.variables = {
+      details_link: {
+        onClick: this._handleViewDetails.bind(this),
+        content: '<span role="button" ' +
+          'class="link" ' +
+          'data-variable="details_link">View details</span>'
+      },
+      requires_cli_link: {
+        content: '<a ' +
+          'href="https://jujucharms.com/docs/stable/reference-install" ' +
+          'target="_blank">Juju CLI client</a>'
+      }
+    };
+    this.state= {
+      content: null,
+      metadata: {}
+    };
+  }
+
+  componentDidMount() {
+    const files = this.props.files;
+
+    // We only want to try and get the usage.md file if there's one there.
+    // Unfortunately, the only place we can easily know whether the file
+    // exists or not is from the entity details page.
+    // The + icon in search results will make an api request for the file
+    // and 404 if it's not there.
+    if (!files || (files && files.indexOf('usage.md') !== -1)) {
+      this.props.getFile(
+        this.props.entityId,
+        'usage.md',
+        this._getUsageCallback.bind(this)
+      );
+    }
+  }
+
+  /**
+    Callback when usage.md is retrieved.
+
+    @param {Object} error Error if there is one.
+    @param {String} usageContents The contents of the file.
+  */
+  _getUsageCallback(error, usageContents) {
+    if (usageContents && usageContents.substring(0, 1) !== '{') {
+      this.setState({
+        metadata: this.parseMarkdownMetadata(usageContents)
+      });
+      const markdown = this.props.marked(
+        this.replaceVars(
+          usageContents
+        )
+      );
+
+      this.setState({
+        content: markdown
+      });
+    }
+  }
+
+  /**
+    Parse header information from markdown files.
+    If the markdown passed in starts with:
+    ---
+    metadata: value
+    ---
+    an object like:
+    {
+      metadata: value
+    }
+    will be assigned to the state.
+
+    @param {String} markdown The raw markdown.
+    @return {Object} Parsed metadata.
+  */
+  parseMarkdownMetadata(markdown) {
+    let parsedMetadata = {};
+    if (markdown.indexOf('---') === 0) {
+      let lineByLine = markdown.split('\n');
+      let metadata = [];
+      lineByLine.shift();
+      while(lineByLine[0] !== '---') {
+        metadata.push(lineByLine.shift());
+      }
+      lineByLine.shift();
+
+      metadata.forEach(option => {
```

Comment:
Maybe we could be more defensive re invalid/not well-formed metadata? The usage.md file is supplied by users and the charm store doesn't do any validation on that, so we should expect anything, and cannot assume, for instance, that the `key: value` pattern is always respected. Maybe check and add notification, so that at least we know what's going on.